# **Análisis espacio-temporal del consumo hídrico en Bogotá (2020–2024) y su relación con la variabilidad climática.**

Este cuaderno documenta y organiza los diccionarios de variables climáticas, meteorológicas y oceánicas utilizados en el estudio, proporcionando una caracterización detallada de cada parámetro en términos de su significado, naturaleza del dato y tipo de variable.
Incluye tanto variables físicas (precipitación, temperatura, humedad, viento, radiación y nubosidad) como indicadores climáticos globales (por ejemplo, el Índice Oceánico del Niño – ONI), así como campos espaciales y temporales (año, mes, localidad y coordenadas geográficas).
Su propósito es servir como referencia conceptual para la interpretación de los datos, facilitar la correcta clasificación de las variables y asegurar coherencia metodológica en los análisis posteriores.

## **Diccionarios**
### **Diccionario variables climáticas**

| PARAMETER | Descripción (inglés) | Traducción / significado en español | Naturaleza del dato | Tipo de variable |
|-----------|----------------------|-------------------------------------|---------------------|------------------|
| IMERG_PRECTOT | IMERG Precipitation Total | Precipitación total (satélite IMERG) | Cuantitativo continuo | Climática – hidrológica |
| PRECTOTCORR | Corrected Total Precipitation | Precipitación total corregida | Cuantitativo continuo | Climática – hidrológica |
| PRECTOTCORR_SUM | Corrected Total Precipitation (sum) | Suma de precipitación corregida | Cuantitativo continuo | Climática – hidrológica |
| QV2M | Specific Humidity at 2m | Humedad específica a 2 metros | Cuantitativo continuo | Atmosférica |
| RH2M | Relative Humidity at 2m | Humedad relativa a 2 metros | Cuantitativo continuo | Atmosférica |
| GWETPROF | Profile Soil Moisture | Humedad del suelo (perfil completo) | Cuantitativo continuo | Hidrológica – suelo |
| GWETROOT | Root Zone Soil Moisture | Humedad del suelo en zona radicular | Cuantitativo continuo | Hidrológica – suelo |
| GWETTOP | Top Soil Moisture | Humedad del suelo superficial | Cuantitativo continuo | Hidrológica – suelo |
| ALLSKY_SFC_LW_DWN | All-sky Surface Longwave Downward | Radiación infrarroja descendente en superficie | Cuantitativo continuo | Radiativa |
| T2M | Temperature at 2m | Temperatura del aire a 2 metros | Cuantitativo continuo | Atmosférica |
| T2MDEW | Dew/Frost Point at 2m | Punto de rocío a 2 metros | Cuantitativo continuo | Atmosférica |
| T2MWET | Wet Bulb Temperature at 2m | Temperatura de bulbo húmedo a 2 metros | Cuantitativo continuo | Atmosférica |
| T2M_MAX | Max Temperature at 2m | Temperatura máxima a 2 metros | Cuantitativo continuo | Atmosférica |
| T2M_MIN | Min Temperature at 2m | Temperatura mínima a 2 metros | Cuantitativo continuo | Atmosférica |
| T2M_RANGE | Temperature Range at 2m | Rango térmico (máx – mín) | Cuantitativo continuo | Atmosférica |
| TS | Earth Skin Temperature | Temperatura de la superficie del suelo | Cuantitativo continuo | Superficie |
| PS | Surface Pressure | Presión atmosférica en superficie | Cuantitativo continuo | Atmosférica |
| WD10M | Wind Direction at 10m | Dirección del viento a 10 m | Cuantitativo continuo (angular) | Atmosférica |
| WD2M | Wind Direction at 2m | Dirección del viento a 2 m | Cuantitativo continuo (angular) | Atmosférica |
| WS10M | Wind Speed at 10m | Velocidad del viento a 10 m | Cuantitativo continuo | Atmosférica |
| WS10M_MAX | Max Wind Speed at 10m | Velocidad máxima del viento a 10 m | Cuantitativo continuo | Atmosférica |
| WS10M_MIN | Min Wind Speed at 10m | Velocidad mínima del viento a 10 m | Cuantitativo continuo | Atmosférica |
| WS10M_RANGE | Wind Speed Range at 10m | Rango de velocidad del viento a 10 m | Cuantitativo continuo | Atmosférica |
| WS2M | Wind Speed at 2m | Velocidad del viento a 2 m | Cuantitativo continuo | Atmosférica |
| WS2M_MAX | Max Wind Speed at 2m | Velocidad máxima del viento a 2 m | Cuantitativo continuo | Atmosférica |
| WS2M_MIN | Min Wind Speed at 2m | Velocidad mínima del viento a 2 m | Cuantitativo continuo | Atmosférica |
| WS2M_RANGE | Wind Speed Range at 2m | Rango de velocidad del viento a 2 m | Cuantitativo continuo | Atmosférica |
| ALLSKY_SFC_PAR_TOT | All-sky Photosynthetically Active Radiation | Radiación fotosintéticamente activa | Cuantitativo continuo | Radiativa – biológica |
| ALLSKY_SFC_SW_DIFF | Diffuse Shortwave Radiation | Radiación solar difusa | Cuantitativo continuo | Radiativa |
| ALLSKY_SFC_SW_DNI | Direct Normal Irradiance | Radiación solar directa normal | Cuantitativo continuo | Radiativa |
| ALLSKY_SFC_SW_DWN | All-sky Shortwave Downward | Radiación solar entrante total | Cuantitativo continuo | Radiativa |
| ALLSKY_SFC_UVA | Surface UVA Irradiance | Radiación ultravioleta A | Cuantitativo continuo | Radiativa |
| ALLSKY_SFC_UVB | Surface UVB Irradiance | Radiación ultravioleta B | Cuantitativo continuo | Radiativa |
| ALLSKY_SFC_UV_INDEX | UV Index | Índice ultravioleta | Cuantitativo continuo | Radiativa – salud |
| ALLSKY_SRF_ALB | Surface Albedo | Albedo superficial (reflectividad) | Cuantitativo continuo | Radiativa |
| CLOUD_AMT | Cloud Amount | Cobertura nubosa | Cuantitativo continuo | Atmosférica |
| CLRSKY_SFC_PAR_TOT | Clear-sky PAR | Radiación fotosintética sin nubes | Cuantitativo continuo | Radiativa |
| CLRSKY_SFC_SW_DWN | Clear-sky SW Downward | Radiación solar sin nubes | Cuantitativo continuo | Radiativa |
| TOA_SW_DWN | Top of Atmosphere SW Downward | Radiación solar en el tope de la atmósfera | Cuantitativo continuo | Radiativa |
| ONI | Oceanic Niño Index | Índice Oceánico del Niño (ENSO) | Cuantitativo continuo | Climática – oceánica |

### **Diccionario variables categoricas**

| Columna | Descripción | Naturaleza del dato | Tipo de variable |
|---------|-------------|--------------------|------------------|
| YEAR | Año de observación | Cuantitativo discreto | Temporal |
| localidad | Nombre de la localidad | Cualitativo nominal | Espacial |
| latitud | Coordenada geográfica (N-S) | Cuantitativo continuo | Espacial |
| longitud | Coordenada geográfica (E-O) | Cuantitativo continuo | Espacial |
| MES | Mes en formato texto (JAN–DEC) | Cualitativo ordinal | Temporal |
| MES_NUM | Mes en número (1–12) | Cuantitativo discreto | Temporal |
| VALOR | Valor observado del parámetro | Cuantitativo continuo | Climático |

## 1. Importación de librerías

En este bloque se importan las librerías necesarias para la lectura y manipulación de archivos:
- `pandas` se utiliza para trabajar con estructuras de datos tipo DataFrame.
- `glob` permite gestionar conjuntos de archivos mediante patrones de búsqueda (aunque en este caso las rutas se definen manualmente).

se importan las librerías necesarias para el procesamiento, análisis y modelado de los datos mediante técnicas de aprendizaje no supervisado:

- **NumPy (`numpy`)**: se utiliza para operaciones matemáticas y manejo eficiente de arreglos numéricos.
- **Seaborn (`seaborn`)** y **Matplotlib (`matplotlib.pyplot`)**: se emplean para la visualización gráfica de los datos y de los resultados del análisis.
- **Warnings (`warnings`)** y **Math (`math`)**: permiten controlar mensajes de advertencia y realizar cálculos matemáticos auxiliares.

En cuanto a las herramientas de aprendizaje automático:

- **`KMeans`, `AgglomerativeClustering`, `DBSCAN`**: algoritmos de clustering utilizados para la segmentación de los datos en grupos con características similares.
- **`GaussianMixture`**: modelo probabilístico de mezcla de gaussianas para clustering basado en distribuciones.
- **`StandardScaler`, `MinMaxScaler`**: métodos de normalización y estandarización de variables numéricas.
- **`OneHotEncoder`, `OrdinalEncoder`**: técnicas de codificación para variables categóricas.
- **`SimpleImputer`**: herramienta para el tratamiento de valores faltantes mediante imputación.
- **`ColumnTransformer`**: permite aplicar distintos procesos de transformación a subconjuntos de columnas.
- **`Pipeline`**: facilita la integración secuencial de etapas de preprocesamiento y modelado.
- **`GridSearchCV`**: se emplea para la optimización de hiperparámetros de los modelos.
- **`PCA` (Análisis de Componentes Principales)** y **`TSNE`**: técnicas de reducción de dimensionalidad utilizadas para visualización y análisis estructural de los datos.
- **`silhouette_score`**: métrica utilizada para evaluar la calidad de los agrupamientos obtenidos.
- **`dendrogram`, `linkage`**: funciones empleadas en la visualización de clustering jerárquico.
- **`joblib.dump`**: se utiliza para guardar modelos entrenados y objetos de procesamiento.
- **`make_scorer`**: permite definir métricas personalizadas para evaluación de modelos.

Este conjunto de librerías proporciona las herramientas necesarias para desarrollar un flujo completo de análisis que incluye preprocesamiento de datos, aplicación de algoritmos de clustering, evaluación de resultados y visualización de patrones.



## 2. Definición de rutas de archivos CSV por localidad

En este bloque se construye una lista (`rutas_csv`) que contiene las rutas completas de los archivos CSV correspondientes a las bases de datos limpias de cada localidad analizada.  
Cada archivo representa la información climática procesada para una localidad específica, lo que permite su posterior integración en una sola base de datos general.



## 3. Definición de la ruta del archivo Excel

En este bloque se define la ruta del archivo Excel que contiene el Índice Oceánico del Niño (ONI) con frecuencia mensual.  
Este archivo se incorpora posteriormente al conjunto de datos para incluir una variable climática global en el análisis.



## 4. Lectura de los archivos CSV

En este bloque se leen todos los archivos CSV definidos en la lista `rutas_csv` y se almacenan en una lista de DataFrames (`lista_df`).  
Se utiliza el parámetro `encoding="latin1"` para asegurar una correcta lectura de caracteres especiales presentes en los nombres de localidades y variables.



## 5. Lectura del archivo Excel

En este bloque se carga el archivo Excel que contiene los datos del ONI en un DataFrame independiente (`df_excel`).  
Este paso permite integrar información oceánica global con las variables meteorológicas locales.



## 6. Integración de todos los conjuntos de datos

En este bloque se añade el DataFrame del archivo Excel a la lista de DataFrames y posteriormente se concatenan todos en un único DataFrame (`df_total`).  
La función `pd.concat()` permite unir los datos por filas, mientras que el parámetro `ignore_index=True` reinicia el índice para garantizar una numeración continua.



## 7. Almacenamiento de la base de datos consolidada

En este bloque se exporta el DataFrame final (`df_total`) a un archivo CSV denominado `base_unida_total.csv`.  
Se utiliza la codificación `utf-8` para garantizar compatibilidad con distintos sistemas y aplicaciones.



## 8. Confirmación del proceso

En este bloque se imprime un mensaje en pantalla indicando que la unión de las bases de datos se realizó correctamente, lo cual permite verificar de forma rápida la finalización exitosa del proceso.

In [115]:
import pandas as pd
import glob

import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
import math
from sklearn.cluster import KMeans,AgglomerativeClustering, DBSCAN
from sklearn.preprocessing import StandardScaler
from scipy.cluster.hierarchy import dendrogram, linkage
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.metrics import silhouette_score
import warnings
import math
from sklearn.preprocessing import StandardScaler, MinMaxScaler,  OneHotEncoder, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from joblib import dump
from sklearn.preprocessing import StandardScaler 
from sklearn.cluster import KMeans
from sklearn.mixture import GaussianMixture

from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.metrics import make_scorer

rutas_csv = [
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_AN.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_BU.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_BO.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_CHA.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_CBO.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_EN.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_FON.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_KEN.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_CAN.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_MAR.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_PUA.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_RAF.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_SAN.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_Santa.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_SUB.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_SUM.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_TEU.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_TUNJ.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_USAQ.csv",
    r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\base_limpia_USM.csv"
]

ruta_excel = r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\oni_mesual.xlsx"

# Leer CSV con encoding latino
lista_df = [pd.read_csv(f, encoding="latin1") for f in rutas_csv]

# Leer Excel aparte
df_excel = pd.read_excel(ruta_excel)

# Unir todo
lista_df.append(df_excel)
df_total = pd.concat(lista_df, ignore_index=True)


print("Base unida correctamente ✅")

Base unida correctamente ✅


## 9. Visualización preliminar del conjunto de datos consolidado

En este bloque se muestran las primeras filas del DataFrame final (`df_total`) mediante el método `head()`.  
Esta visualización inicial permite verificar la correcta integración de los datos provenientes de las distintas localidades y del archivo adicional, así como comprobar la estructura general de las columnas y el formato de los valores.

Este paso constituye una validación exploratoria básica previa al preprocesamiento y al análisis mediante técnicas de clustering.

In [116]:
df_total.head()

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
0,IMERG_PRECTOT,2020,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
1,IMERG_PRECTOT,2021,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
2,IMERG_PRECTOT,2022,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
3,IMERG_PRECTOT,2023,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
4,IMERG_PRECTOT,2024,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207


## 9. Identificación de la dimensión del conjunto de datos

En este bloque se obtiene el tamaño del DataFrame consolidado (`df_total`) mediante el atributo `shape`, el cual devuelve:
- el número de registros (filas), y  
- el número de variables (columnas).

Esta información permite conocer la magnitud del conjunto de datos y resulta fundamental para planificar las etapas posteriores de preprocesamiento, modelado y análisis computacional.

In [117]:
# cantidad de variables y registros
df_total.shape

(4005, 18)

## 10. Inspección estructural y muestreo del conjunto de datos

En este bloque se realiza una exploración más detallada de la estructura del DataFrame consolidado (`df_total`) mediante distintas visualizaciones:

- Se muestran los **primeros tres registros** (`head(3)`) para observar el inicio del conjunto de datos.
- Se visualizan los **últimos tres registros** (`tail(3)`) para verificar la consistencia al final del DataFrame.
- Se listan las **columnas** del DataFrame (`columns`) para identificar las variables disponibles.
- Se muestran los **índices** (`index`) para comprobar la numeración de los registros.
- Se consultan los **tipos de datos** de cada columna (`dtypes`) para validar su naturaleza (numérica, categórica, etc.).
- Se extrae una **muestra aleatoria de tres registros** (`sample(3)`) con el fin de observar valores representativos sin sesgo por posición.

Este conjunto de operaciones permite verificar la coherencia interna del DataFrame y facilita la identificación temprana de posibles problemas de formato o tipado de variables antes del preprocesamiento.

In [118]:

display("primeros 3 registros",df_total.head(3))
display("3 últimos registros",df_total.tail(3))
display("Columnas:",df_total.columns)
display("Índices:",df_total.index)
display("Tipos de registros en el DataFrame",df_total.dtypes)
display("Muestra aleatoria de 3 registros",df_total.sample(3))

'primeros 3 registros'

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
0,IMERG_PRECTOT,2020,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
1,IMERG_PRECTOT,2021,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
2,IMERG_PRECTOT,2022,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207


'3 últimos registros'

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
4002,ONI,2022,-0.97,-0.93,-0.99,-1.06,-0.99,-0.85,-0.81,-0.91,-1.01,-0.99,-0.92,-0.83,NaN,Usme,NaN,NaN
4003,ONI,2023,-0.68,-0.43,-0.15,0.16,0.48,0.77,1.07,1.32,1.56,1.78,1.92,1.95,NaN,Usme,NaN,NaN
4004,ONI,2024,1.78,1.48,1.14,0.71,0.39,0.15,0.04,-0.11,-0.21,-0.26,-0.37,-0.53,NaN,Usme,NaN,NaN


'Columnas:'

Index(['PARAMETER', 'YEAR', 'JAN', 'FEB', 'MAR', 'APR', 'MAY', 'JUN', 'JUL',
       'AUG', 'SEP', 'OCT', 'NOV', 'DEC', 'ANN', 'localidad', 'latitud',
       'longitud'],
      dtype='object')

'Índices:'

RangeIndex(start=0, stop=4005, step=1)

'Tipos de registros en el DataFrame'

PARAMETER     object
YEAR           int64
JAN          float64
FEB          float64
MAR          float64
APR          float64
MAY          float64
JUN          float64
JUL          float64
AUG          float64
SEP          float64
OCT          float64
NOV          float64
DEC          float64
ANN          float64
localidad     object
latitud      float64
longitud     float64
dtype: object

'Muestra aleatoria de 3 registros'

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
1527,ALLSKY_SFC_UVB,2022,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,0.04,Kennedy,4.62598,-74.10040
1985,GWETTOP,2020,0.67,0.64,0.65,0.66,0.65,0.73,0.71,0.68,0.67,0.60,0.61,0.57,0.65,Puente Aranda,4.63111,-74.11198
1523,ALLSKY_SFC_UVA,2023,1.19,1.23,1.01,1.04,1.23,1.28,1.25,1.23,1.23,1.19,1.16,1.13,1.18,Kennedy,4.62598,-74.10040


## 11. Resumen general de la estructura del conjunto de datos

En este bloque se utiliza el método `info()` para obtener un resumen técnico del DataFrame consolidado (`df_total`).  
Este resumen incluye:

- El número total de registros.
- El número de columnas.
- El nombre de cada variable.
- El tipo de dato asociado a cada columna.
- La cantidad de valores no nulos por variable.
- El uso aproximado de memoria.

Esta información es fundamental para evaluar la calidad inicial del conjunto de datos, identificar variables con valores faltantes y verificar que los tipos de datos sean coherentes con el análisis estadístico y los modelos de clustering que se aplicarán posteriormente.

In [119]:
df_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4005 entries, 0 to 4004
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   PARAMETER  4005 non-null   object 
 1   YEAR       4005 non-null   int64  
 2   JAN        4005 non-null   float64
 3   FEB        4005 non-null   float64
 4   MAR        4005 non-null   float64
 5   APR        4005 non-null   float64
 6   MAY        4005 non-null   float64
 7   JUN        4005 non-null   float64
 8   JUL        4005 non-null   float64
 9   AUG        4005 non-null   float64
 10  SEP        4005 non-null   float64
 11  OCT        4005 non-null   float64
 12  NOV        4005 non-null   float64
 13  DEC        4005 non-null   float64
 14  ANN        3905 non-null   float64
 15  localidad  4005 non-null   object 
 16  latitud    3905 non-null   float64
 17  longitud   3905 non-null   float64
dtypes: float64(15), int64(1), object(2)
memory usage: 563.3+ KB


## 12. Análisis y eliminación de filas duplicadas

En este bloque se evalúa la existencia de registros duplicados en el conjunto de datos consolidado (`df_total`).  
Inicialmente, se imprime el tamaño del DataFrame antes de eliminar duplicados con el fin de contar con una referencia del número total de filas.

Posteriormente, se aplica el método `drop_duplicates()` para eliminar posibles registros repetidos, generando un nuevo DataFrame sin duplicados.  
A continuación, se muestra nuevamente el tamaño del conjunto de datos, lo que permite comparar ambos valores y verificar si existían filas duplicadas.

Finalmente, se visualizan las primeras filas del DataFrame para confirmar que la estructura se mantiene consistente tras el proceso de depuración.

Este procedimiento contribuye a mejorar la calidad de los datos y a evitar sesgos en los análisis y modelos posteriores.

In [120]:
#Análisis de filas duplicadas en la DB, calculamos las filas antes de eliminar las repetidas
print(f'Tamaño del set antes de eliminar las filas repetidas: {df_total.shape}')
 
#Eliminamos filas duplicadas de la DB
df = df_total.drop_duplicates()
#Impresión del tamaño del set luego de eliminar las duplicadas
 
print(f'Tamaño del set después de eliminar las filas repetidas: {df_total.shape}')
#Se evidencia que no se encuentran valores duplicados
df_total.head()

Tamaño del set antes de eliminar las filas repetidas: (4005, 18)
Tamaño del set después de eliminar las filas repetidas: (4005, 18)


,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
0,IMERG_PRECTOT,2020,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
1,IMERG_PRECTOT,2021,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
2,IMERG_PRECTOT,2022,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
3,IMERG_PRECTOT,2023,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207
4,IMERG_PRECTOT,2024,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,Antonio Nariño,4.60971,-74.08207


## 13. Identificación de valores nulos en el conjunto de datos

En este bloque se realiza un análisis de valores faltantes en el DataFrame consolidado (`df_total`).  
Mediante la función `isnull().sum()`, se calcula el número total de valores nulos por cada variable, lo que permite:

- Detectar columnas con información incompleta.
- Evaluar la magnitud del problema de datos faltantes.
- Definir estrategias posteriores de tratamiento, como imputación de valores o eliminación de variables/registros.

Este paso es fundamental para garantizar la calidad del conjunto de datos antes de aplicar procesos de normalización, reducción de dimensionalidad y algoritmos de clustering.

In [121]:
# Identificar valores nulos
display(df_total.isnull().sum())

PARAMETER      0
YEAR           0
JAN            0
FEB            0
MAR            0
APR            0
MAY            0
JUN            0
JUL            0
AUG            0
SEP            0
OCT            0
NOV            0
DEC            0
ANN          100
localidad      0
latitud      100
longitud     100
dtype: int64

## 14. Análisis estadístico descriptivo de las variables numéricas

En este bloque se utiliza el método `describe()` para obtener un resumen estadístico de las variables cuantitativas del DataFrame consolidado (`df_total`).  
Este resumen incluye, para cada variable numérica:

- El número de observaciones válidas (count).
- La media (mean).
- La desviación estándar (std).
- El valor mínimo (min).
- Los cuartiles (25%, 50% y 75%).
- El valor máximo (max).

Este análisis permite identificar el rango de variación de las variables, detectar posibles valores atípicos y comprender la distribución general de los datos, lo cual resulta esencial para orientar las etapas posteriores de preprocesamiento y modelado mediante técnicas de clustering.

In [122]:
df_total.describe()

,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,latitud,longitud
count,4005.00000,4005.000000,4005.000000,4005.000000,4005.000000,4005.000000,4005.000000,4005.00000,4005.000000,4005.000000,4005.000000,4005.000000,4005.000000,3905.000000,3905.000000,3905.000000
mean,2022.00000,-5.982891,-4.986287,-4.281179,0.029501,1.158909,0.718649,-0.94603,-2.178794,-1.742607,-1.882487,-2.144135,-5.646197,40.355706,4.611792,-74.102881
std,1.41439,161.298389,161.551959,162.467063,165.706200,167.953154,166.696393,164.82362,163.491901,163.622863,164.211218,165.852979,161.474470,337.657365,0.112061,0.043307
min,2020.00000,-999.000000,-999.000000,-999.000000,-999.000000,-999.000000,-999.000000,-999.00000,-999.000000,-999.000000,-999.000000,-999.000000,-999.000000,-999.000000,4.200000,-74.216680
25%,2021.00000,1.250000,1.210000,1.050000,1.040000,1.180000,1.230000,1.24000,1.230000,1.280000,1.240000,1.120000,1.130000,1.230000,4.590280,-74.111980
50%,2022.00000,8.910000,8.710000,7.760000,8.060000,8.430000,8.530000,8.53000,8.560000,8.790000,8.820000,8.070000,8.570000,9.180000,4.609470,-74.090350
75%,2023.00000,20.350000,22.240000,22.030000,21.190000,21.510000,20.550000,20.63000,21.180000,22.400000,22.280000,19.260000,21.140000,24.990000,4.670400,-74.076060
max,2024.00000,245.830000,145.090000,359.700000,336.710000,552.790000,453.980000,402.73000,268.080000,223.910000,359.900000,359.500000,234.640000,3049.710000,4.759400,-74.043380


## 15. Identificación de valores únicos por variable

En este bloque se recorre cada columna del DataFrame (`df_total`) para calcular:
- La cantidad de valores únicos (`nunique()`).
- El conjunto de valores distintos presentes en cada variable (`unique()`).

Este análisis permite detectar posibles inconsistencias, errores de codificación y variables categóricas, así como comprender la diversidad de datos almacenados.

In [123]:
for col in df_total.columns:
    print(f"\nColumna: {col}")
    print("Cantidad de únicos:", df_total[col].nunique())
    print("Valores únicos:", df_total[col].unique())


Columna: PARAMETER
Cantidad de únicos: 40
Valores únicos: ['IMERG_PRECTOT' 'PRECTOTCORR' 'PRECTOTCORR_SUM' 'QV2M' 'RH2M' 'GWETPROF'
 'GWETROOT' 'GWETTOP' 'ALLSKY_SFC_LW_DWN' 'T2M' 'T2MDEW' 'T2MWET'
 'T2M_MAX' 'T2M_MIN' 'T2M_RANGE' 'TS' 'PS' 'WD10M' 'WD2M' 'WS10M'
 'WS10M_MAX' 'WS10M_MIN' 'WS10M_RANGE' 'WS2M' 'WS2M_MAX' 'WS2M_MIN'
 'WS2M_RANGE' 'ALLSKY_SFC_PAR_TOT' 'ALLSKY_SFC_SW_DIFF'
 'ALLSKY_SFC_SW_DNI' 'ALLSKY_SFC_SW_DWN' 'ALLSKY_SFC_UVA' 'ALLSKY_SFC_UVB'
 'ALLSKY_SFC_UV_INDEX' 'ALLSKY_SRF_ALB' 'CLOUD_AMT' 'CLRSKY_SFC_PAR_TOT'
 'CLRSKY_SFC_SW_DWN' 'TOA_SW_DWN' 'ONI']

Columna: YEAR
Cantidad de únicos: 5
Valores únicos: [2020 2021 2022 2023 2024]

Columna: JAN
Cantidad de únicos: 477
Valores únicos: [-9.9900e+02  7.6000e-01  3.3900e+00  1.8500e+00  5.8500e+00  6.2000e-01
  2.3510e+01  1.0514e+02  5.7490e+01  1.8132e+02  1.9180e+01  1.2950e+01
  1.1860e+01  1.1980e+01  1.2330e+01  1.2420e+01  7.9680e+01  7.3390e+01
  7.1830e+01  8.2160e+01  7.3000e-01  5.1000e-01  8.0000e-01  6.8000e

## 16. Asignación y corrección de coordenadas geográficas por localidad

En este bloque se define un diccionario que asocia cada localidad con sus coordenadas geográficas (latitud y longitud).  
Posteriormente, se crean columnas temporales para asignar dichas coordenadas a partir del nombre de la localidad.

Las coordenadas faltantes (`NaN`) se rellenan usando la información del diccionario, y finalmente se eliminan las columnas auxiliares.  
Este procedimiento garantiza la completitud y consistencia de la información espacial del conjunto de datos.


In [124]:
import pandas as pd
import numpy as np

# Diccionario de coordenadas
coords = {
    "Antonio Nariño": (4.60971, -74.08207),
    "Barrios Unidos": (4.69890, -74.07075),
    "Bosa": (4.58734, -74.21668),
    "Chapinero": (4.67040, -74.04338),
    "Ciudad Bolívar": (4.53904, -74.14732),
    "Engativá": (4.70179, -74.10088),
    "Fontibón": (4.71049, -74.14500),
    "Kennedy": (4.62598, -74.10040),
    "La Candelaria": (4.59806, -74.07606),
    "Los Mártires": (4.60055, -74.07481),
    "Puente Aranda": (4.63111, -74.11198),
    "Rafael Uribe Uribe": (4.59028, -74.08473),
    "San Cristóbal": (4.58843, -74.07080),
    "Santa Fe": (4.60947, -74.08135),
    "Suba": (4.75940, -74.09035),
    "Sumapaz": (4.20000, -74.20000),
    "Teusaquillo": (4.65942, -74.09919),
    "Tunjuelito": (4.59069, -74.09628),
    "Usaquén": (4.73197, -74.05163),
    "Usme": (4.53286, -74.11451)
}

# Crear columnas auxiliares desde el diccionario
df["lat_temp"] = df["localidad"].map(lambda x: coords.get(x, (np.nan, np.nan))[0])
df["lon_temp"] = df["localidad"].map(lambda x: coords.get(x, (np.nan, np.nan))[1])

# Rellenar solo donde haya NaN
df["latitud"] = df["latitud"].fillna(df["lat_temp"])
df["longitud"] = df["longitud"].fillna(df["lon_temp"])

# Borrar columnas temporales
df.drop(columns=["lat_temp", "lon_temp"], inplace=True)

C:\Users\Xiomara\AppData\Local\Temp\ipykernel_16760\1173773949.py:29: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["lat_temp"] = df["localidad"].map(lambda x: coords.get(x, (np.nan, np.nan))[0])
C:\Users\Xiomara\AppData\Local\Temp\ipykernel_16760\1173773949.py:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["lon_temp"] = df["localidad"].map(lambda x: coords.get(x, (np.nan, np.nan))[1])
C:\Users\Xiomara\AppData\Local\Temp\ipykernel_16760\1173773949.py:33: SettingWithCopyWarning: 
A value is trying t

## 17. Normalización y limpieza de nombres de localidades

En este bloque se define una función que estandariza los nombres de las localidades mediante:
- Corrección de posibles errores de codificación.
- Eliminación de tildes y caracteres especiales.
- Conversión del texto a mayúsculas.
- Eliminación de espacios innecesarios.

Este proceso asegura homogeneidad en la variable `localidad` y evita duplicidades causadas por diferencias ortográficas.

In [125]:
import unicodedata

def limpiar_localidad(x):
    if pd.isna(x):
        return x
    x = str(x)

    # intentar corregir mala codificación
    try:
        x = x.encode("latin1").decode("utf-8")
    except:
        pass  # si ya está bien, no hace nada

    # quitar tildes
    x = unicodedata.normalize("NFKD", x).encode("ascii", "ignore").decode("ascii")

    # pasar a mayúsculas
    x = x.upper().strip()

    return x

df["localidad"] = df["localidad"].apply(limpiar_localidad)

df["localidad"].unique()

C:\Users\Xiomara\AppData\Local\Temp\ipykernel_16760\2277785687.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["localidad"] = df["localidad"].apply(limpiar_localidad)


array(['ANTONIO NARINO', 'BARRIOS UNIDOS', 'BOSA', 'CHAPINERO',
       'CIUDAD BOLIVAR', 'ENGATIVA', 'FONTIBON', 'KENNEDY',
       'LA CANDELARIA', 'LOS MARTIRES', 'PUENTE ARANDA',
       'RAFAEL URIBE URIBE', 'SAN CRISTOBAL', 'SANTA FE', 'SUBA',
       'SUMAPAZ', 'TEUSAQUILLO', 'TUNJUELITO', 'USAQUEN', 'USME'],
      dtype=object)

## 18. Visualización preliminar del DataFrame procesado

En este bloque se muestran las primeras filas del DataFrame (`df`) para verificar que los procesos de limpieza y estandarización se hayan aplicado correctamente.

In [126]:
df.head()

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
0,IMERG_PRECTOT,2020,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,ANTONIO NARINO,4.60971,-74.08207
1,IMERG_PRECTOT,2021,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,ANTONIO NARINO,4.60971,-74.08207
2,IMERG_PRECTOT,2022,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,ANTONIO NARINO,4.60971,-74.08207
3,IMERG_PRECTOT,2023,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,ANTONIO NARINO,4.60971,-74.08207
4,IMERG_PRECTOT,2024,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,-999.0,ANTONIO NARINO,4.60971,-74.08207


## 19. Verificación de valores faltantes en coordenadas geográficas

En este bloque se contabilizan los valores nulos en las columnas `latitud` y `longitud`.  
Este paso permite confirmar que la asignación de coordenadas se realizó correctamente y que no existen registros sin información espacial.


In [127]:
df[["latitud", "longitud"]].isna().sum()

latitud     0
longitud    0
dtype: int64

## 20. Extracción de registros correspondientes al índice ONI

En este bloque se filtran los registros cuyo parámetro corresponde al índice ONI (`PARAMETER == "ONI"`).  
Esta selección permite analizar de forma específica la variable climática oceánica y verificar su correcta integración en el conjunto de datos.


In [128]:
df_oni = df[df["PARAMETER"] == "ONI"]
df_oni.head(10)

,PARAMETER,YEAR,JAN,FEB,MAR,APR,MAY,JUN,JUL,AUG,SEP,OCT,NOV,DEC,ANN,localidad,latitud,longitud
3905,ONI,2020,0.50,0.48,0.40,0.19,-0.08,-0.30,-0.41,-0.57,-0.89,-1.17,-1.27,-1.19,NaN,ANTONIO NARINO,4.60971,-74.08207
3906,ONI,2021,-1.05,-0.93,-0.84,-0.66,-0.48,-0.38,-0.40,-0.49,-0.67,-0.81,-0.98,-0.98,NaN,ANTONIO NARINO,4.60971,-74.08207
3907,ONI,2022,-0.97,-0.93,-0.99,-1.06,-0.99,-0.85,-0.81,-0.91,-1.01,-0.99,-0.92,-0.83,NaN,ANTONIO NARINO,4.60971,-74.08207
3908,ONI,2023,-0.68,-0.43,-0.15,0.16,0.48,0.77,1.07,1.32,1.56,1.78,1.92,1.95,NaN,ANTONIO NARINO,4.60971,-74.08207
3909,ONI,2024,1.78,1.48,1.14,0.71,0.39,0.15,0.04,-0.11,-0.21,-0.26,-0.37,-0.53,NaN,ANTONIO NARINO,4.60971,-74.08207
3910,ONI,2020,0.50,0.48,0.40,0.19,-0.08,-0.30,-0.41,-0.57,-0.89,-1.17,-1.27,-1.19,NaN,BARRIOS UNIDOS,4.69890,-74.07075
3911,ONI,2021,-1.05,-0.93,-0.84,-0.66,-0.48,-0.38,-0.40,-0.49,-0.67,-0.81,-0.98,-0.98,NaN,BARRIOS UNIDOS,4.69890,-74.07075
3912,ONI,2022,-0.97,-0.93,-0.99,-1.06,-0.99,-0.85,-0.81,-0.91,-1.01,-0.99,-0.92,-0.83,NaN,BARRIOS UNIDOS,4.69890,-74.07075
3913,ONI,2023,-0.68,-0.43,-0.15,0.16,0.48,0.77,1.07,1.32,1.56,1.78,1.92,1.95,NaN,BARRIOS UNIDOS,4.69890,-74.07075
3914,ONI,2024,1.78,1.48,1.14,0.71,0.39,0.15,0.04,-0.11,-0.21,-0.26,-0.37,-0.53,NaN,BARRIOS UNIDOS,4.69890,-74.07075


## 21. Eliminación de la columna anual agregada

En este bloque se elimina la columna `ANN`, la cual representa un valor anual agregado.  
Su eliminación es necesaria para evitar redundancias, dado que el análisis se realiza a nivel mensual.


In [129]:
df = df.drop(columns=["ANN"])

## 22. Transformación del conjunto de datos a formato largo (long format)

En este bloque se reorganiza el DataFrame desde formato ancho (meses como columnas) a formato largo mediante la función `melt()`.  
Las columnas de meses (JAN–DEC) se transforman en una sola variable (`MES`) y sus valores asociados en la columna `VALOR`.

Esta transformación es fundamental para facilitar análisis temporales, visualizaciones y posteriores procesos de reestructuración.


In [130]:
df_param_long = df.melt(
    id_vars=["PARAMETER", "YEAR", "localidad", "latitud", "longitud"],
    value_vars=["JAN","FEB","MAR","APR","MAY","JUN","JUL","AUG","SEP","OCT","NOV","DEC"],
    var_name="MES",
    value_name="VALOR"
)

## 23. Visualización de registros en formato largo

En este bloque se muestran los primeros registros del DataFrame en formato largo (`df_param_long`) con el fin de validar la correcta conversión desde el formato ancho.

In [131]:
df_param_long.head(30)

,PARAMETER,YEAR,localidad,latitud,longitud,MES,VALOR
0,IMERG_PRECTOT,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00
1,IMERG_PRECTOT,2021,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00
2,IMERG_PRECTOT,2022,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00
3,IMERG_PRECTOT,2023,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00
4,IMERG_PRECTOT,2024,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00
5,PRECTOTCORR,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,0.76
6,PRECTOTCORR,2021,ANTONIO NARINO,4.60971,-74.08207,JAN,3.39
7,PRECTOTCORR,2022,ANTONIO NARINO,4.60971,-74.08207,JAN,1.85
8,PRECTOTCORR,2023,ANTONIO NARINO,4.60971,-74.08207,JAN,5.85
9,PRECTOTCORR,2024,ANTONIO NARINO,4.60971,-74.08207,JAN,0.62


## 24. Codificación numérica del mes

En este bloque se crea un diccionario que asocia cada mes en formato texto (JAN–DEC) con su correspondiente valor numérico (1–12).  
Posteriormente, se genera la columna `MES_NUM`, lo que facilita la ordenación cronológica y el análisis temporal.

In [132]:
meses = {
    "JAN": 1, "FEB": 2, "MAR": 3, "APR": 4,
    "MAY": 5, "JUN": 6, "JUL": 7, "AUG": 8,
    "SEP": 9, "OCT": 10, "NOV": 11, "DEC": 12
}

# 'MES' fue creada en df_param_long al hacer melt, por eso mapear sobre df_param_long
df_param_long["MES_NUM"] = df_param_long["MES"].map(meses)

## 25. Análisis de valores únicos en el DataFrame en formato largo

En este bloque se examina cada columna del DataFrame `df_param_long` para identificar:
- El número de valores únicos.
- Los valores distintos contenidos en cada variable.

Este análisis permite verificar la integridad de las variables transformadas y detectar posibles anomalías tras la conversión.

In [133]:
for col in df_param_long:
    print(f"\nColumna: {col}")
    print("Cantidad de únicos:", df_param_long[col].nunique())
    print("Valores únicos:", df_param_long[col].unique())


Columna: PARAMETER
Cantidad de únicos: 40
Valores únicos: ['IMERG_PRECTOT' 'PRECTOTCORR' 'PRECTOTCORR_SUM' 'QV2M' 'RH2M' 'GWETPROF'
 'GWETROOT' 'GWETTOP' 'ALLSKY_SFC_LW_DWN' 'T2M' 'T2MDEW' 'T2MWET'
 'T2M_MAX' 'T2M_MIN' 'T2M_RANGE' 'TS' 'PS' 'WD10M' 'WD2M' 'WS10M'
 'WS10M_MAX' 'WS10M_MIN' 'WS10M_RANGE' 'WS2M' 'WS2M_MAX' 'WS2M_MIN'
 'WS2M_RANGE' 'ALLSKY_SFC_PAR_TOT' 'ALLSKY_SFC_SW_DIFF'
 'ALLSKY_SFC_SW_DNI' 'ALLSKY_SFC_SW_DWN' 'ALLSKY_SFC_UVA' 'ALLSKY_SFC_UVB'
 'ALLSKY_SFC_UV_INDEX' 'ALLSKY_SRF_ALB' 'CLOUD_AMT' 'CLRSKY_SFC_PAR_TOT'
 'CLRSKY_SFC_SW_DWN' 'TOA_SW_DWN' 'ONI']

Columna: YEAR
Cantidad de únicos: 5
Valores únicos: [2020 2021 2022 2023 2024]

Columna: localidad
Cantidad de únicos: 20
Valores únicos: ['ANTONIO NARINO' 'BARRIOS UNIDOS' 'BOSA' 'CHAPINERO' 'CIUDAD BOLIVAR'
 'ENGATIVA' 'FONTIBON' 'KENNEDY' 'LA CANDELARIA' 'LOS MARTIRES'
 'PUENTE ARANDA' 'RAFAEL URIBE URIBE' 'SAN CRISTOBAL' 'SANTA FE' 'SUBA'
 'SUMAPAZ' 'TEUSAQUILLO' 'TUNJUELITO' 'USAQUEN' 'USME']

Columna: latitud
Ca

## 26. Inspección visual del DataFrame en formato largo

En este bloque se muestran las primeras filas del DataFrame `df_param_long` para confirmar la coherencia de la estructura y de los valores almacenados.


In [134]:
df_param_long.head(10)

,PARAMETER,YEAR,localidad,latitud,longitud,MES,VALOR,MES_NUM
0,IMERG_PRECTOT,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00,1
1,IMERG_PRECTOT,2021,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00,1
2,IMERG_PRECTOT,2022,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00,1
3,IMERG_PRECTOT,2023,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00,1
4,IMERG_PRECTOT,2024,ANTONIO NARINO,4.60971,-74.08207,JAN,-999.00,1
5,PRECTOTCORR,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,0.76,1
6,PRECTOTCORR,2021,ANTONIO NARINO,4.60971,-74.08207,JAN,3.39,1
7,PRECTOTCORR,2022,ANTONIO NARINO,4.60971,-74.08207,JAN,1.85,1
8,PRECTOTCORR,2023,ANTONIO NARINO,4.60971,-74.08207,JAN,5.85,1
9,PRECTOTCORR,2024,ANTONIO NARINO,4.60971,-74.08207,JAN,0.62,1


## 27. Reestructuración del DataFrame a formato ancho por parámetro

En este bloque se reorganiza el conjunto de datos mediante una tabla pivote (`pivot_table`), donde:
- Cada fila representa una combinación única de año, localidad y mes.
- Cada columna corresponde a un parámetro climático.
- Los valores corresponden a las observaciones registradas.

Este formato es más adecuado para análisis multivariados y técnicas de aprendizaje automático.

In [135]:
df_param_pivot = df_param_long.pivot_table(
    index=["YEAR", "localidad", "latitud", "longitud", "MES", "MES_NUM"],
    columns="PARAMETER",
    values="VALOR"
).reset_index()

df_param_pivot.head(20)

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,ANTONIO NARINO,4.60971,-74.08207,APR,4,32.62,7.91,8.64,7.69,...,100.5,103.3,1.41,4.47,0.14,1.93,0.89,3.29,0.05,1.50
1,2020,ANTONIO NARINO,4.60971,-74.08207,AUG,8,32.02,7.89,8.11,8.93,...,125.1,127.7,1.77,4.73,0.13,2.08,1.12,3.46,0.07,1.68
2,2020,ANTONIO NARINO,4.60971,-74.08207,DEC,12,32.08,8.24,8.02,12.64,...,55.3,51.3,1.41,4.41,0.06,2.08,0.87,3.21,0.04,1.62
3,2020,ANTONIO NARINO,4.60971,-74.08207,FEB,2,31.78,9.07,8.00,13.87,...,85.7,86.6,2.18,6.09,0.08,2.25,1.36,4.34,0.10,1.87
4,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,1,31.45,8.89,7.59,15.18,...,70.7,71.0,2.37,7.21,0.33,2.37,1.50,5.13,0.20,2.02
5,2020,ANTONIO NARINO,4.60971,-74.08207,JUL,7,31.94,7.92,8.57,7.89,...,123.8,125.3,1.82,4.80,0.12,2.18,1.13,3.44,0.05,1.71
6,2020,ANTONIO NARINO,4.60971,-74.08207,JUN,6,31.87,7.96,8.75,7.69,...,114.7,115.8,1.99,6.05,0.06,2.42,1.24,4.26,0.05,1.89
7,2020,ANTONIO NARINO,4.60971,-74.08207,MAR,3,32.74,7.76,8.35,7.72,...,55.8,53.2,1.69,5.28,0.10,2.47,1.05,3.84,0.07,1.94
8,2020,ANTONIO NARINO,4.60971,-74.08207,MAY,5,32.56,7.73,8.87,6.34,...,116.4,118.3,1.84,4.77,0.08,2.06,1.14,3.40,0.04,1.66
9,2020,ANTONIO NARINO,4.60971,-74.08207,NOV,11,32.21,7.67,8.51,8.18,...,15.6,359.5,1.29,4.71,0.04,1.94,0.81,3.44,0.03,1.52


## 28. Ordenamiento temporal y espacial del conjunto de datos

En este bloque se ordenan los registros del DataFrame pivoteado (`df_param_pivot`) por localidad, año y mes numérico.  
Este ordenamiento garantiza una estructura cronológica coherente para análisis posteriores.

In [136]:


# ordenar por localidad y mes
df_param_pivot = df_param_pivot.sort_values(
    by=["localidad", "YEAR", "MES_NUM"]
).reset_index(drop=True)

df_param_pivot.head(20)

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,1,31.45,8.89,7.59,15.18,...,70.7,71.0,2.37,7.21,0.33,2.37,1.50,5.13,0.20,2.02
1,2020,ANTONIO NARINO,4.60971,-74.08207,FEB,2,31.78,9.07,8.00,13.87,...,85.7,86.6,2.18,6.09,0.08,2.25,1.36,4.34,0.10,1.87
2,2020,ANTONIO NARINO,4.60971,-74.08207,MAR,3,32.74,7.76,8.35,7.72,...,55.8,53.2,1.69,5.28,0.10,2.47,1.05,3.84,0.07,1.94
3,2020,ANTONIO NARINO,4.60971,-74.08207,APR,4,32.62,7.91,8.64,7.69,...,100.5,103.3,1.41,4.47,0.14,1.93,0.89,3.29,0.05,1.50
4,2020,ANTONIO NARINO,4.60971,-74.08207,MAY,5,32.56,7.73,8.87,6.34,...,116.4,118.3,1.84,4.77,0.08,2.06,1.14,3.40,0.04,1.66
5,2020,ANTONIO NARINO,4.60971,-74.08207,JUN,6,31.87,7.96,8.75,7.69,...,114.7,115.8,1.99,6.05,0.06,2.42,1.24,4.26,0.05,1.89
6,2020,ANTONIO NARINO,4.60971,-74.08207,JUL,7,31.94,7.92,8.57,7.89,...,123.8,125.3,1.82,4.80,0.12,2.18,1.13,3.44,0.05,1.71
7,2020,ANTONIO NARINO,4.60971,-74.08207,AUG,8,32.02,7.89,8.11,8.93,...,125.1,127.7,1.77,4.73,0.13,2.08,1.12,3.46,0.07,1.68
8,2020,ANTONIO NARINO,4.60971,-74.08207,SEP,9,32.20,8.79,8.64,11.16,...,123.0,128.0,1.82,4.73,0.04,2.13,1.13,3.40,0.03,1.72
9,2020,ANTONIO NARINO,4.60971,-74.08207,OCT,10,32.19,8.28,8.64,9.97,...,117.4,121.3,2.02,4.57,0.06,2.08,1.27,3.35,0.03,1.75


## 29. Identificación de dimensiones del conjunto de datos final

En este bloque se obtiene el tamaño del DataFrame (`df_param_pivot`) mediante `shape`, lo que permite conocer:
- El número total de observaciones.
- El número de variables incluidas.

In [137]:
# cantidad de variables y registros
df_param_pivot.shape

(1200, 46)

## 30. Exploración estructural del conjunto de datos final

En este bloque se realiza una exploración integral del DataFrame final (`df_param_pivot`) que incluye:
- Visualización de los primeros y últimos registros.
- Inspección de columnas e índices.
- Verificación de tipos de datos.
- Generación de una muestra aleatoria.
- Resumen técnico mediante `info()`.
- Identificación de valores nulos por variable.
- Análisis de valores únicos por columna.

Este conjunto de operaciones permite validar la calidad, coherencia y estructura final del conjunto de datos antes de aplicar técnicas de normalización, reducción de dimensionalidad y algoritmos de clustering.

In [138]:
display("primeros 3 registros",df_param_pivot.head(3))
display("3 últimos registros",df_param_pivot.tail(3))
display("Columnas:",df_param_pivot.columns)
display("Índices:",df_param_pivot.index)
display("Tipos de registros en el DataFrame",df_param_pivot.dtypes)
display("Muestra aleatoria de 3 registros",df_param_pivot.sample(3))

'primeros 3 registros'

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,1,31.45,8.89,7.59,15.18,...,70.7,71.0,2.37,7.21,0.33,2.37,1.50,5.13,0.20,2.02
1,2020,ANTONIO NARINO,4.60971,-74.08207,FEB,2,31.78,9.07,8.00,13.87,...,85.7,86.6,2.18,6.09,0.08,2.25,1.36,4.34,0.10,1.87
2,2020,ANTONIO NARINO,4.60971,-74.08207,MAR,3,32.74,7.76,8.35,7.72,...,55.8,53.2,1.69,5.28,0.10,2.47,1.05,3.84,0.07,1.94


'3 últimos registros'

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
1197,2024,USME,4.53286,-74.11451,OCT,10,33.33,9.06,8.86,11.14,...,108.8,112.1,1.81,7.23,0.03,2.21,1.14,5.15,0.04,1.73
1198,2024,USME,4.53286,-74.11451,NOV,11,33.50,7.48,7.86,8.58,...,42.9,38.1,1.57,4.37,0.09,2.11,0.98,3.10,0.03,1.64
1199,2024,USME,4.53286,-74.11451,DEC,12,33.62,7.58,7.31,11.02,...,59.4,57.1,1.47,3.50,0.07,1.92,0.91,2.61,0.04,1.49


'Columnas:'

Index(['YEAR', 'localidad', 'latitud', 'longitud', 'MES', 'MES_NUM',
       'ALLSKY_SFC_LW_DWN', 'ALLSKY_SFC_PAR_TOT', 'ALLSKY_SFC_SW_DIFF',
       'ALLSKY_SFC_SW_DNI', 'ALLSKY_SFC_SW_DWN', 'ALLSKY_SFC_UVA',
       'ALLSKY_SFC_UVB', 'ALLSKY_SFC_UV_INDEX', 'ALLSKY_SRF_ALB', 'CLOUD_AMT',
       'CLRSKY_SFC_PAR_TOT', 'CLRSKY_SFC_SW_DWN', 'GWETPROF', 'GWETROOT',
       'GWETTOP', 'IMERG_PRECTOT', 'ONI', 'PRECTOTCORR', 'PRECTOTCORR_SUM',
       'PS', 'QV2M', 'RH2M', 'T2M', 'T2MDEW', 'T2MWET', 'T2M_MAX', 'T2M_MIN',
       'T2M_RANGE', 'TOA_SW_DWN', 'TS', 'WD10M', 'WD2M', 'WS10M', 'WS10M_MAX',
       'WS10M_MIN', 'WS10M_RANGE', 'WS2M', 'WS2M_MAX', 'WS2M_MIN',
       'WS2M_RANGE'],
      dtype='object', name='PARAMETER')

'Índices:'

RangeIndex(start=0, stop=1200, step=1)

'Tipos de registros en el DataFrame'

PARAMETER
YEAR                     int64
localidad               object
latitud                float64
longitud               float64
MES                     object
MES_NUM                  int64
ALLSKY_SFC_LW_DWN      float64
ALLSKY_SFC_PAR_TOT     float64
ALLSKY_SFC_SW_DIFF     float64
ALLSKY_SFC_SW_DNI      float64
ALLSKY_SFC_SW_DWN      float64
ALLSKY_SFC_UVA         float64
ALLSKY_SFC_UVB         float64
ALLSKY_SFC_UV_INDEX    float64
ALLSKY_SRF_ALB         float64
CLOUD_AMT              float64
CLRSKY_SFC_PAR_TOT     float64
CLRSKY_SFC_SW_DWN      float64
GWETPROF               float64
GWETROOT               float64
GWETTOP                float64
IMERG_PRECTOT          float64
ONI                    float64
PRECTOTCORR            float64
PRECTOTCORR_SUM        float64
PS                     float64
QV2M                   float64
RH2M                   float64
T2M                    float64
T2MDEW                 float64
T2MWET                 float64
T2M_MAX                float6

'Muestra aleatoria de 3 registros'

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
690,2022,RAFAEL URIBE URIBE,4.59028,-74.08473,JUL,7,32.91,7.70,8.00,8.63,...,118.0,120.0,1.82,5.42,0.17,2.17,1.14,3.90,0.09,1.68
482,2020,LA CANDELARIA,4.59806,-74.07606,MAR,3,32.74,7.76,8.35,7.72,...,55.8,53.2,1.69,5.28,0.10,2.47,1.05,3.84,0.07,1.94
233,2024,CHAPINERO,4.67040,-74.04338,JUN,6,33.44,7.77,8.53,7.36,...,106.1,108.1,2.26,5.53,0.30,3.54,1.45,3.57,0.19,2.40


## 32. Inspección de la estructura del DataFrame  
Se revisa la información general del conjunto de datos para conocer el número de columnas, el tipo de dato de cada una y la presencia de valores no nulos.

## 33. Identificación de valores nulos  
Se contabilizan los valores faltantes por columna con el fin de detectar variables con datos incompletos o problemáticos.

## 34. Análisis de valores únicos por variable  
Se examinan los valores distintos de cada columna para identificar posibles valores atípicos o códigos especiales utilizados para representar datos faltantes.

## 35. Eliminación de la variable IMERG_PRECTOT  
Se elimina la columna *IMERG_PRECTOT* debido a que presenta el valor **-999**, el cual corresponde a un código utilizado para representar datos faltantes y no una medición real de precipitación.

## 36. Verificación de las columnas restantes  
Se revisa la lista de columnas del DataFrame para confirmar que la variable *IMERG_PRECTOT* fue eliminada correctamente.

## 37. Visualización de los primeros registros  
Se muestran las primeras filas del conjunto de datos resultante para validar la correcta depuración y la estructura final del DataFrame.

In [139]:
df_param_pivot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 46 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   YEAR                 1200 non-null   int64  
 1   localidad            1200 non-null   object 
 2   latitud              1200 non-null   float64
 3   longitud             1200 non-null   float64
 4   MES                  1200 non-null   object 
 5   MES_NUM              1200 non-null   int64  
 6   ALLSKY_SFC_LW_DWN    1200 non-null   float64
 7   ALLSKY_SFC_PAR_TOT   1200 non-null   float64
 8   ALLSKY_SFC_SW_DIFF   1200 non-null   float64
 9   ALLSKY_SFC_SW_DNI    1200 non-null   float64
 10  ALLSKY_SFC_SW_DWN    1200 non-null   float64
 11  ALLSKY_SFC_UVA       1200 non-null   float64
 12  ALLSKY_SFC_UVB       1200 non-null   float64
 13  ALLSKY_SFC_UV_INDEX  1200 non-null   float64
 14  ALLSKY_SRF_ALB       1200 non-null   float64
 15  CLOUD_AMT            1200 non-null   f

In [140]:
# Identificar valores nulos
display(df_param_pivot.isnull().sum())

PARAMETER
YEAR                   0
localidad              0
latitud                0
longitud               0
MES                    0
MES_NUM                0
ALLSKY_SFC_LW_DWN      0
ALLSKY_SFC_PAR_TOT     0
ALLSKY_SFC_SW_DIFF     0
ALLSKY_SFC_SW_DNI      0
ALLSKY_SFC_SW_DWN      0
ALLSKY_SFC_UVA         0
ALLSKY_SFC_UVB         0
ALLSKY_SFC_UV_INDEX    0
ALLSKY_SRF_ALB         0
CLOUD_AMT              0
CLRSKY_SFC_PAR_TOT     0
CLRSKY_SFC_SW_DWN      0
GWETPROF               0
GWETROOT               0
GWETTOP                0
IMERG_PRECTOT          0
ONI                    0
PRECTOTCORR            0
PRECTOTCORR_SUM        0
PS                     0
QV2M                   0
RH2M                   0
T2M                    0
T2MDEW                 0
T2MWET                 0
T2M_MAX                0
T2M_MIN                0
T2M_RANGE              0
TOA_SW_DWN             0
TS                     0
WD10M                  0
WD2M                   0
WS10M                  0
WS10M_MAX      

In [141]:
for col in df_param_pivot.columns:
    print(f"\nColumna: {col}")
    print("Cantidad de únicos:", df_param_pivot[col].nunique())
    print("Valores únicos:", df_param_pivot[col].unique())


Columna: YEAR
Cantidad de únicos: 5
Valores únicos: [2020 2021 2022 2023 2024]

Columna: localidad
Cantidad de únicos: 20
Valores únicos: ['ANTONIO NARINO' 'BARRIOS UNIDOS' 'BOSA' 'CHAPINERO' 'CIUDAD BOLIVAR'
 'ENGATIVA' 'FONTIBON' 'KENNEDY' 'LA CANDELARIA' 'LOS MARTIRES'
 'PUENTE ARANDA' 'RAFAEL URIBE URIBE' 'SAN CRISTOBAL' 'SANTA FE' 'SUBA'
 'SUMAPAZ' 'TEUSAQUILLO' 'TUNJUELITO' 'USAQUEN' 'USME']

Columna: latitud
Cantidad de únicos: 20
Valores únicos: [4.60971 4.6989  4.58734 4.6704  4.53904 4.70179 4.71049 4.62598 4.59806
 4.60055 4.63111 4.59028 4.58843 4.60947 4.7594  4.2     4.65942 4.59069
 4.73197 4.53286]

Columna: longitud
Cantidad de únicos: 20
Valores únicos: [-74.08207 -74.07075 -74.21668 -74.04338 -74.14732 -74.10088 -74.145
 -74.1004  -74.07606 -74.07481 -74.11198 -74.08473 -74.0708  -74.08135
 -74.09035 -74.2     -74.09919 -74.09628 -74.05163 -74.11451]

Columna: MES
Cantidad de únicos: 12
Valores únicos: ['JAN' 'FEB' 'MAR' 'APR' 'MAY' 'JUN' 'JUL' 'AUG' 'SEP' 'OCT' 'NO

In [142]:
df_param_pivot = df_param_pivot.drop(columns=["IMERG_PRECTOT"])

In [143]:
df_param_pivot.columns.tolist()

['YEAR',
 'localidad',
 'latitud',
 'longitud',
 'MES',
 'MES_NUM',
 'ALLSKY_SFC_LW_DWN',
 'ALLSKY_SFC_PAR_TOT',
 'ALLSKY_SFC_SW_DIFF',
 'ALLSKY_SFC_SW_DNI',
 'ALLSKY_SFC_SW_DWN',
 'ALLSKY_SFC_UVA',
 'ALLSKY_SFC_UVB',
 'ALLSKY_SFC_UV_INDEX',
 'ALLSKY_SRF_ALB',
 'CLOUD_AMT',
 'CLRSKY_SFC_PAR_TOT',
 'CLRSKY_SFC_SW_DWN',
 'GWETPROF',
 'GWETROOT',
 'GWETTOP',
 'ONI',
 'PRECTOTCORR',
 'PRECTOTCORR_SUM',
 'PS',
 'QV2M',
 'RH2M',
 'T2M',
 'T2MDEW',
 'T2MWET',
 'T2M_MAX',
 'T2M_MIN',
 'T2M_RANGE',
 'TOA_SW_DWN',
 'TS',
 'WD10M',
 'WD2M',
 'WS10M',
 'WS10M_MAX',
 'WS10M_MIN',
 'WS10M_RANGE',
 'WS2M',
 'WS2M_MAX',
 'WS2M_MIN',
 'WS2M_RANGE']

In [144]:
df_param_pivot.head()

PARAMETER,YEAR,localidad,latitud,longitud,MES,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,ANTONIO NARINO,4.60971,-74.08207,JAN,1,31.45,8.89,7.59,15.18,...,70.7,71.0,2.37,7.21,0.33,2.37,1.50,5.13,0.20,2.02
1,2020,ANTONIO NARINO,4.60971,-74.08207,FEB,2,31.78,9.07,8.00,13.87,...,85.7,86.6,2.18,6.09,0.08,2.25,1.36,4.34,0.10,1.87
2,2020,ANTONIO NARINO,4.60971,-74.08207,MAR,3,32.74,7.76,8.35,7.72,...,55.8,53.2,1.69,5.28,0.10,2.47,1.05,3.84,0.07,1.94
3,2020,ANTONIO NARINO,4.60971,-74.08207,APR,4,32.62,7.91,8.64,7.69,...,100.5,103.3,1.41,4.47,0.14,1.93,0.89,3.29,0.05,1.50
4,2020,ANTONIO NARINO,4.60971,-74.08207,MAY,5,32.56,7.73,8.87,6.34,...,116.4,118.3,1.84,4.77,0.08,2.06,1.14,3.40,0.04,1.66


In [145]:
df_param_pivot.shape

(1200, 45)

## 38. Exportación del DataFrame climático limpio
Se guarda el conjunto de datos estructurado y depurado en formato Excel en la ruta local del proyecto para asegurar el almacenamiento de la base de datos climática primaria.

## 39. Eliminación de variables de granularidad mensual
Se remueven las columnas correspondientes al mes ("MES" y "MES_NUM") con el fin de preparar el DataFrame para transformaciones y agregaciones a escalas temporales de mayor nivel.

## 40. Agregación de datos a nivel anual
Se agrupan los registros climáticos por año, localidad y coordenadas geográficas para calcular el promedio anual de cada variable meteorológica mediante funciones numéricas.

## 41. Almacenamiento de la base histórica anual
El DataFrame con los promedios anuales calculados se exporta a un archivo de Excel para su posterior análisis o visualización macroecológica.

## 42. Verificación de la estructura consolidada anual
Se realiza una inspección visual de las primeras filas del nuevo conjunto de datos anual para validar la correcta ejecución de la agrupación y el estado de las variables.



In [146]:
df_param_pivot.to_excel(r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\bd_var_clima_limpia.xlsx", index=False)

In [147]:
# eliminar columnas de mes
df_param_pivot1 = df_param_pivot.drop(columns=["MES", "MES_NUM"])

# agrupar a nivel anual
df_anual = df_param_pivot.groupby(
    ["YEAR", "localidad", "latitud", "longitud"],
    as_index=False
).mean(numeric_only=True)

# guardar resultado
df_anual.to_excel("base_anual.xlsx", index=False)

df_anual.head()

PARAMETER,YEAR,localidad,latitud,longitud,MES_NUM,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,ALLSKY_SFC_SW_DWN,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,ANTONIO NARINO,4.60971,-74.08207,6.5,32.138333,8.175833,8.390833,9.771667,17.188333,...,92.000,121.775000,1.800833,5.151667,0.103333,2.165833,1.125833,3.713333,0.063333,1.74
1,2020,BARRIOS UNIDOS,4.69890,-74.07075,6.5,32.138333,8.175833,8.390833,9.771667,17.188333,...,92.000,121.775000,1.800833,5.151667,0.103333,2.165833,1.125833,3.713333,0.063333,1.74
2,2020,BOSA,4.58734,-74.21668,6.5,32.138333,8.175833,8.390833,9.771667,17.188333,...,92.000,121.775000,1.800833,5.151667,0.103333,2.165833,1.125833,3.713333,0.063333,1.74
3,2020,CHAPINERO,4.67040,-74.04338,6.5,32.138333,8.175833,8.390833,9.771667,17.188333,...,103.425,106.366667,1.973333,5.375833,0.092500,3.170833,1.225000,3.517500,0.053333,2.14
4,2020,CIUDAD BOLIVAR,4.53904,-74.14732,6.5,32.138333,8.175833,8.390833,9.771667,17.188333,...,92.000,121.775000,1.800833,5.151667,0.103333,2.165833,1.125833,3.713333,0.063333,1.74


## 43. Construcción del indicador bimestral
Se aplica una operación aritmética sobre el número del mes para discretizar el tiempo y crear una nueva variable categórica y numérica que identifica el bimestre del año.

## 44. Definición de criterios para la agrupación bimestral
Se estructuran los arreglos con las variables espaciotemporales de agrupación y las variables climáticas de tipo numérico, excluyendo los índices mensuales que ya no son requeridos.

## 45. Agregación de variables meteorológicas por bimestre
Se ejecuta la agrupación por año, localidad y bimestre, aplicando el promedio matemático a todas las métricas climáticas para estandarizar la escala temporal de la información.

## 46. Visualización del DataFrame bimestral resultante
Se examinan los primeros registros de la base meteorológica bimestral para corroborar que la segmentación temporal cumpla con el formato y los valores esperados.

## 47. Almacenamiento de la base climática bimestral
Se genera y guarda un archivo en formato CSV en el repositorio del proyecto que contiene la información meteorológica procesada de manera bimestral.

## 48. Inspección de dimensiones de la matriz bimestral
Se evalúa el número total de filas y columnas del conjunto de datos climáticos bimestral para asegurar que el volumen de datos coincida con los criterios metodológicos establecidos.

## 49. Carga de los registros históricos de consumo
Se realiza la lectura e importación de la base de datos socioeconómica o de consumo en formato Excel para iniciar la fase de vinculación con las variables del entorno.

## 50. Exploración inicial de la base de consumo
Se despliegan las primeras filas del DataFrame de consumo cargado para comprender su estructura lógica, los nombres de los encabezados y la naturaleza de sus registros.

## 51. Listado analítico de variables de consumo
Se extrae y visualiza la lista completa de las columnas del DataFrame de consumo con el objetivo de identificar las llaves primarias necesarias para el cruce de información.

## 52. Homologación del identificador temporal de consumo
Se renombra la variable original de tiempo ("v") a "BIMESTRE" para unificar la nomenclatura técnica con el resto de las matrices de datos del proyecto.

## 53. Mapeo y codificación numérica de bimestres
Se transforman las etiquetas textuales del consumo (ej. "v1", "v2") en valores numéricos consecutivos (del 1 al 6) mediante un diccionario de mapeo para estandarizar la variable temporal.

## 54. Auditoría de frecuencias por intervalo bimestral
Se calcula la distribución y cantidad de registros por cada bimestre con el fin de detectar desbalances, pérdidas de datos o inconsistencias en la codificación.

## 55. Extracción de localidades en la base de consumo
Se listan los valores únicos de la columna de ubicaciones espaciales en el conjunto de consumo para identificar la nomenclatura regional utilizada.

In [148]:
# Crear columna de bimestre (solo una vez)
df_param_pivot["BIMESTRE"] = ((df_param_pivot["MES_NUM"] - 1) // 2) + 1

cols_group = ["YEAR", "localidad", "latitud", "longitud", "BIMESTRE"]

cols_numeric = df_param_pivot.select_dtypes(include="number").columns
cols_numeric = cols_numeric.drop("MES_NUM")

df_bimestral = (
    df_param_pivot
    .groupby(cols_group, as_index=False)[cols_numeric]
    .mean()
)

df_bimestral.head()

PARAMETER,localidad,YEAR,latitud,longitud,ALLSKY_SFC_LW_DWN,ALLSKY_SFC_PAR_TOT,ALLSKY_SFC_SW_DIFF,ALLSKY_SFC_SW_DNI,ALLSKY_SFC_SW_DWN,ALLSKY_SFC_UVA,...,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE,BIMESTRE
0,ANTONIO NARINO,2020.0,4.60971,-74.08207,31.615,8.980,7.795,14.525,19.135,1.290,...,78.80,2.275,6.650,0.205,2.310,1.430,4.735,0.150,1.945,1.0
1,ANTONIO NARINO,2020.0,4.60971,-74.08207,32.680,7.835,8.495,7.705,16.660,1.095,...,78.25,1.550,4.875,0.120,2.200,0.970,3.565,0.060,1.720,2.0
2,ANTONIO NARINO,2020.0,4.60971,-74.08207,32.215,7.845,8.810,7.015,16.035,1.230,...,117.05,1.915,5.410,0.070,2.240,1.190,3.830,0.045,1.775,3.0
3,ANTONIO NARINO,2020.0,4.60971,-74.08207,31.980,7.905,8.340,8.410,16.460,1.180,...,126.50,1.795,4.765,0.125,2.130,1.125,3.450,0.060,1.695,4.0
4,ANTONIO NARINO,2020.0,4.60971,-74.08207,32.195,8.535,8.640,10.565,18.020,1.215,...,124.65,1.920,4.650,0.050,2.105,1.200,3.375,0.030,1.735,5.0


In [149]:
df_bimestral.to_csv("C:\\Users\\Xiomara\\Documents\\Projects\\Tesis\\cd_tesis\\Par_meteo\\processed_db\\base_bimestral_var_clim.csv", index=False)

In [150]:
df_bimestral.shape

(600, 44)

In [151]:
df_consumo = pd.read_excel(r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\DB_CONSUMO.xlsx")

df_consumo.head()

,año,localidad,v,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor
0,2020,USAQUEN,v1,4653292,213354,10.905097,212731,215097
1,2020,CHAPINERO,v1,2281601,93362,12.219109,93016,93684
2,2020,SANTA FE,v1,843631,33789,12.483811,33524,34703
3,2020,SAN CRISTOBAL,v1,1771678,80676,10.980205,80311,81072
4,2020,USME,v1,1570282,78240,10.035033,77876,79159


In [152]:
df_consumo.columns.tolist()

['año',
 'localidad',
 'v',
 'metros_cubicos',
 'facturas',
 'consumo_promedio',
 'suscriptores_facturados',
 'suscriptores_con_medidor']

In [153]:
df_consumo = df_consumo.rename(columns={"v": "BIMESTRE"})

df_consumo["BIMESTRE"] = (
    df_consumo["BIMESTRE"]
    .map({
        "v1": 1,
        "v2": 2,
        "v3": 3,
        "v4": 4,
        "v5": 5,
        "v6": 6
    })
)

In [154]:
df_consumo["BIMESTRE"].value_counts()

BIMESTRE
1    100
2    100
3    100
4    100
5    100
6    100
Name: count, dtype: int64

In [155]:
df_consumo["localidad"].unique()

array(['USAQUEN', 'CHAPINERO', 'SANTA FE', 'SAN CRISTOBAL', 'USME',
       'TUNJUELITO', 'BOSA', 'KENNEDY', 'FONTIBON', 'ENGATIVA', 'SUBA',
       'BARRIOS UNIDOS', 'TEUSAQUILLO', 'LOS MARTIRES', 'ANTONIO NARIÑO',
       'PUENTE ARANDA', 'CANDELARIA', 'RAFAEL URIBE', 'CIUDAD BOLIVAR',
       'OTROS'], dtype=object)

## 56. Extracción de localidades en la base meteorológica
Se obtienen los nombres únicos de las localidades en el DataFrame climático para comparar la consistencia ortográfica e identificar discrepancias textuales entre ambas fuentes.

## 57. Inspección preliminar de consistencia en consumo
Se revisa de nuevo la cabecera de la base de consumo para certificar el éxito de las transformaciones de nombres y tipos de datos efectuadas previamente.

## 58. Diseño del diccionario de homologación geográfica
Se construye una estructura de diccionario detallada para corregir variaciones en tildes, caracteres especiales o nombres de localidades, garantizando la paridad textual exacta.

## 59. Aplicación de la corrección de nombres de localidades
Se reemplazan los nombres de las zonas geográficas en el DataFrame de consumo utilizando el diccionario de homologación para erradicar errores de unión por discrepancias de texto.

## 60. Estandarización de la variable temporal anual
Se renombra el encabezado del año ("YEAR") en la base bimestral meteorológica a minúsculas ("año") para que coincida exactamente con la llave de combinación del consumo.

## 61. Integración de bases de datos mediante combinación interna
Se ejecuta un Merge de tipo Inner combinando la información de consumo y clima bajo las tres llaves espaciotemporales unificadas: año, localidad y bimestre.

## 62. Verificación dimensional del DataFrame integrado
Se consultan las dimensiones finales de la matriz combinada para comprobar cuántos registros cumplieron exitosamente con los criterios de intersección de la combinación.

## 63. Análisis exploratorio y descripción de variables consolidadas
Se genera un ciclo de inspección que recorre cada columna de la base combinada para imprimir la cantidad de casos únicos y el catálogo de valores existentes, permitiendo diagnosticar la calidad del dato.

## 64. Respaldo de la base de datos combinada
Se exporta el DataFrame integrado (consumo + clima) a un archivo de Excel en la ruta del proyecto como un hito de almacenamiento de la información unificada.

## 65. Resumen estadístico descriptivo
Se calculan las métricas estadísticas clave (media, desviación estándar, mínimos, máximos y percentiles) para evaluar la distribución general de todas las variables del sistema.

## 66. Diseño conceptual del diccionario de traducción de variables
Se estructura un mapa de equivalencias exhaustivo (rename_dict) para renombrar las variables técnicas y códigos de agencias científicas (NASA, IMERG, ONI) a términos formales y descriptivos en español (Precipitación, Humedad, Temperatura, Radiación, etc.).

## 67. Reordenamiento lógico de los componentes espaciotemporales
Se define y aplica un orden de columnas jerárquico para el DataFrame final (df_PBI), posicionando primero los índices de control (año, bimestre, localidad, latitud, longitud) seguidos del bloque de variables científicas.

## 68. Almacenamiento definitivo de la base de datos
Se guarda la matriz estructurada, traducida y ordenada en la ruta final de datos procesados, quedando completamente lista para su consumo en herramientas de visualización (Power BI / Tableau) o modelamiento estadístico avanzado.

In [156]:
df_param_pivot["localidad"].unique()

array(['ANTONIO NARINO', 'BARRIOS UNIDOS', 'BOSA', 'CHAPINERO',
       'CIUDAD BOLIVAR', 'ENGATIVA', 'FONTIBON', 'KENNEDY',
       'LA CANDELARIA', 'LOS MARTIRES', 'PUENTE ARANDA',
       'RAFAEL URIBE URIBE', 'SAN CRISTOBAL', 'SANTA FE', 'SUBA',
       'SUMAPAZ', 'TEUSAQUILLO', 'TUNJUELITO', 'USAQUEN', 'USME'],
      dtype=object)

In [157]:
df_consumo.head()

,año,localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor
0,2020,USAQUEN,1,4653292,213354,10.905097,212731,215097
1,2020,CHAPINERO,1,2281601,93362,12.219109,93016,93684
2,2020,SANTA FE,1,843631,33789,12.483811,33524,34703
3,2020,SAN CRISTOBAL,1,1771678,80676,10.980205,80311,81072
4,2020,USME,1,1570282,78240,10.035033,77876,79159


In [158]:
dic_localidades = {
    "ANTONIO NARIÑO": "ANTONIO NARINO",
    "CANDELARIA": "LA CANDELARIA",
    "RAFAEL URIBE": "RAFAEL URIBE URIBE",
    "OTROS": "SUMAPAZ"  # si quieres homologar OTROS con SUMAPAZ (opcional)
}

df_consumo["localidad"] = df_consumo["localidad"].replace(dic_localidades)

In [159]:
df_bimestral = df_bimestral.rename(columns={"YEAR": "año"})

In [160]:
df_bime_total = pd.merge(
    df_consumo,
    df_bimestral,
    on=["año", "localidad", "BIMESTRE"],
    how="inner"   # solo lo que coincide
)

In [161]:
df_bime_total.head()


,año,localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,latitud,longitud,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,USAQUEN,1,4653292,213354,10.905097,212731,215097,4.73197,-74.05163,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
1,2020,CHAPINERO,1,2281601,93362,12.219109,93016,93684,4.67040,-74.04338,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
2,2020,SANTA FE,1,843631,33789,12.483811,33524,34703,4.60947,-74.08135,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
3,2020,SAN CRISTOBAL,1,1771678,80676,10.980205,80311,81072,4.58843,-74.07080,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
4,2020,USME,1,1570282,78240,10.035033,77876,79159,4.53286,-74.11451,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945


In [162]:
df_bime_total.shape

(600, 49)

In [163]:
df_bime_total.head()

,año,localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,latitud,longitud,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,USAQUEN,1,4653292,213354,10.905097,212731,215097,4.73197,-74.05163,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
1,2020,CHAPINERO,1,2281601,93362,12.219109,93016,93684,4.67040,-74.04338,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
2,2020,SANTA FE,1,843631,33789,12.483811,33524,34703,4.60947,-74.08135,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
3,2020,SAN CRISTOBAL,1,1771678,80676,10.980205,80311,81072,4.58843,-74.07080,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
4,2020,USME,1,1570282,78240,10.035033,77876,79159,4.53286,-74.11451,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945


In [164]:
for col in df_bime_total.columns:
    print(f"\nColumna: {col}")
    print("Cantidad de únicos:", df_bime_total[col].nunique())
    print("Valores únicos:", df_bime_total[col].unique())


Columna: año
Cantidad de únicos: 5
Valores únicos: [2020 2021 2022 2023 2024]

Columna: localidad
Cantidad de únicos: 20
Valores únicos: ['USAQUEN' 'CHAPINERO' 'SANTA FE' 'SAN CRISTOBAL' 'USME' 'TUNJUELITO'
 'BOSA' 'KENNEDY' 'FONTIBON' 'ENGATIVA' 'SUBA' 'BARRIOS UNIDOS'
 'TEUSAQUILLO' 'LOS MARTIRES' 'ANTONIO NARINO' 'PUENTE ARANDA'
 'LA CANDELARIA' 'RAFAEL URIBE URIBE' 'CIUDAD BOLIVAR' 'SUMAPAZ']

Columna: BIMESTRE
Cantidad de únicos: 6
Valores únicos: [1 2 3 4 5 6]

Columna: metros_cubicos
Cantidad de únicos: 600
Valores únicos: [4653292 2281601  843631 1771678 1570282 1030285 3165511 5689236 3178670
 4655902 7038503 1223620 1600057  700018  573064 2372568  263918 1850487
 2752105    1348 4511113 1816336  723282 1863170 1638489 1016725 3235177
 5683447 3122338 4559865 7249852 1095998 1345004  700307  579003 2424312
  197050 1940096 2842300    1306 4627114 2011988  740141 1844269 1691932
 1029659 3218949 5691281 3215730 4542685 7131427 1196802 1424969  706939
  574170 2426110  211465 

In [165]:
ruta = r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\\df_bime_total.xlsx"
df_bime_total.to_excel(ruta, index=False)

In [166]:
df_bime_total.describe()

,año,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,latitud,longitud,ALLSKY_SFC_LW_DWN,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
count,600.000000,600.00000,6.000000e+02,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,...,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000,600.000000
mean,2022.000000,3.50000,2.328651e+06,108915.265000,25.072726,108472.926667,108985.600000,4.611795,-74.102908,32.811667,...,97.739167,108.578833,1.824900,5.259683,0.108025,2.364183,1.152133,3.755183,0.061200,1.825350
std,1.415394,1.70925,1.852242e+06,93368.555012,188.558880,93278.295223,93751.273397,0.112212,0.043358,0.646119,...,29.652937,36.090971,0.363374,0.938463,0.075478,0.377715,0.240015,0.637335,0.043631,0.238564
min,2020.000000,1.00000,3.800000e+01,1.000000,5.080645,0.000000,1.000000,4.200000,-74.216680,31.615000,...,10.150000,7.050000,1.320000,3.585000,0.015000,1.820000,0.810000,2.485000,0.005000,1.430000
25%,2021.000000,2.00000,8.634895e+05,35675.000000,10.119474,35357.500000,36121.750000,4.589817,-74.112612,32.195000,...,82.700000,86.150000,1.535000,4.650000,0.070000,2.155000,0.970000,3.340000,0.038750,1.665000
50%,2022.000000,3.50000,1.842930e+06,82151.500000,10.868146,81786.500000,81923.000000,4.609590,-74.093315,32.767500,...,102.050000,111.800000,1.705000,5.000000,0.095000,2.220000,1.070000,3.640000,0.050000,1.735000
75%,2023.000000,5.00000,3.175161e+06,149652.750000,11.819164,142017.750000,139349.750000,4.677525,-74.075748,33.495000,...,121.000000,126.400000,2.030000,5.850000,0.120000,2.415000,1.285000,4.170000,0.070000,1.945000
max,2024.000000,6.00000,7.319036e+06,368794.000000,3158.666667,368407.000000,369739.000000,4.759400,-74.043380,34.355000,...,256.250000,311.500000,3.745000,9.195000,0.830000,4.135000,2.460000,6.395000,0.440000,2.965000


In [167]:
# Identificar valores nulos
display(df_bime_total.isnull().sum())

año                         0
localidad                   0
BIMESTRE                    0
metros_cubicos              0
facturas                    0
consumo_promedio            0
suscriptores_facturados     0
suscriptores_con_medidor    0
latitud                     0
longitud                    0
ALLSKY_SFC_LW_DWN           0
ALLSKY_SFC_PAR_TOT          0
ALLSKY_SFC_SW_DIFF          0
ALLSKY_SFC_SW_DNI           0
ALLSKY_SFC_SW_DWN           0
ALLSKY_SFC_UVA              0
ALLSKY_SFC_UVB              0
ALLSKY_SFC_UV_INDEX         0
ALLSKY_SRF_ALB              0
CLOUD_AMT                   0
CLRSKY_SFC_PAR_TOT          0
CLRSKY_SFC_SW_DWN           0
GWETPROF                    0
GWETROOT                    0
GWETTOP                     0
ONI                         0
PRECTOTCORR                 0
PRECTOTCORR_SUM             0
PS                          0
QV2M                        0
RH2M                        0
T2M                         0
T2MDEW                      0
T2MWET    

In [168]:
df_bime_total[df_bime_total['suscriptores_facturados'].isnull()]

,año,localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,latitud,longitud,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE


In [169]:
rename_dict = {
    # Temporales y espaciales
    "YEAR": "Año",
    "localidad": "Localidad",
    "latitud": "Latitud",
    "longitud": "Longitud",
    "MES": "Mes",
    "MES_NUM": "Mes_Numero",
    "VALOR": "Valor",

    # Precipitación
    "IMERG_PRECTOT": "Precipitacion_Satelital",
    "PRECTOTCORR": "Precipitacion_Corregida",
    "PRECTOTCORR_SUM": "Precipitacion_Corregida_Acumulada",

    # Humedad
    "QV2M": "Humedad_Especifica_2m",
    "RH2M": "Humedad_Relativa_2m",

    # Suelo
    "GWETPROF": "Humedad_Suelo_Perfil",
    "GWETROOT": "Humedad_Suelo_Raiz",
    "GWETTOP": "Humedad_Suelo_Superficial",

    # Temperatura
    "T2M": "Temperatura_2m",
    "T2MDEW": "Punto_Rocio_2m",
    "T2MWET": "Temperatura_Bulbo_Humedo",
    "T2M_MAX": "Temperatura_Maxima",
    "T2M_MIN": "Temperatura_Minima",
    "T2M_RANGE": "Rango_Temperatura",
    "TS": "Temperatura_Suelo",

    # Presión
    "PS": "Presion_Superficie",

    # Viento
    "WD10M": "Direccion_Viento_10m",
    "WD2M": "Direccion_Viento_2m",
    "WS10M": "Velocidad_Viento_10m",
    "WS10M_MAX": "Velocidad_Viento_Max_10m",
    "WS10M_MIN": "Velocidad_Viento_Min_10m",
    "WS10M_RANGE": "Rango_Viento_10m",
    "WS2M": "Velocidad_Viento_2m",
    "WS2M_MAX": "Velocidad_Viento_Max_2m",
    "WS2M_MIN": "Velocidad_Viento_Min_2m",
    "WS2M_RANGE": "Rango_Viento_2m",

    # Radiación
    "ALLSKY_SFC_LW_DWN": "Radiacion_Infrarroja",
    "ALLSKY_SFC_PAR_TOT": "Radiacion_Fotosintetica",
    "ALLSKY_SFC_SW_DIFF": "Radiacion_Solar_Difusa",
    "ALLSKY_SFC_SW_DNI": "Radiacion_Solar_Directa",
    "ALLSKY_SFC_SW_DWN": "Radiacion_Solar_Total",
    "ALLSKY_SFC_UVA": "Radiacion_UVA",
    "ALLSKY_SFC_UVB": "Radiacion_UVB",
    "ALLSKY_SFC_UV_INDEX": "Indice_UV",
    "ALLSKY_SRF_ALB": "Albedo",

    # Nubes
    "CLOUD_AMT": "Cobertura_Nubosa",

    # Cielo despejado
    "CLRSKY_SFC_PAR_TOT": "Radiacion_Fotosintetica_Despejado",
    "CLRSKY_SFC_SW_DWN": "Radiacion_Solar_Despejado",

    # Atmósfera
    "TOA_SW_DWN": "Radiacion_Tope_Atmosfera",

    # ENSO
    "ONI": "Indice_Nino_ONI"
}

In [170]:
df_bime_total.head()

,año,localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,latitud,longitud,...,WD10M,WD2M,WS10M,WS10M_MAX,WS10M_MIN,WS10M_RANGE,WS2M,WS2M_MAX,WS2M_MIN,WS2M_RANGE
0,2020,USAQUEN,1,4653292,213354,10.905097,212731,215097,4.73197,-74.05163,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
1,2020,CHAPINERO,1,2281601,93362,12.219109,93016,93684,4.67040,-74.04338,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
2,2020,SANTA FE,1,843631,33789,12.483811,33524,34703,4.60947,-74.08135,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
3,2020,SAN CRISTOBAL,1,1771678,80676,10.980205,80311,81072,4.58843,-74.07080,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
4,2020,USME,1,1570282,78240,10.035033,77876,79159,4.53286,-74.11451,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945


In [171]:
df_PBI = df_bime_total.rename(columns=rename_dict)

In [173]:
df_PBI.head()

,año,Localidad,BIMESTRE,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,Latitud,Longitud,...,Direccion_Viento_10m,Direccion_Viento_2m,Velocidad_Viento_10m,Velocidad_Viento_Max_10m,Velocidad_Viento_Min_10m,Rango_Viento_10m,Velocidad_Viento_2m,Velocidad_Viento_Max_2m,Velocidad_Viento_Min_2m,Rango_Viento_2m
0,2020,USAQUEN,1,4653292,213354,10.905097,212731,215097,4.73197,-74.05163,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
1,2020,CHAPINERO,1,2281601,93362,12.219109,93016,93684,4.67040,-74.04338,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
2,2020,SANTA FE,1,843631,33789,12.483811,33524,34703,4.60947,-74.08135,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
3,2020,SAN CRISTOBAL,1,1771678,80676,10.980205,80311,81072,4.58843,-74.07080,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
4,2020,USME,1,1570282,78240,10.035033,77876,79159,4.53286,-74.11451,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945


In [176]:
column_order = [
    "año", "BIMESTRE", "Localidad", "Latitud", "Longitud"
] + [col for col in df_PBI.columns if col not in ["año", "BIMESTRE", "Localidad", "Latitud", "Longitud"]]

df_PBI = df_PBI[column_order]

In [177]:
df_PBI.head()

,año,BIMESTRE,Localidad,Latitud,Longitud,metros_cubicos,facturas,consumo_promedio,suscriptores_facturados,suscriptores_con_medidor,...,Direccion_Viento_10m,Direccion_Viento_2m,Velocidad_Viento_10m,Velocidad_Viento_Max_10m,Velocidad_Viento_Min_10m,Rango_Viento_10m,Velocidad_Viento_2m,Velocidad_Viento_Max_2m,Velocidad_Viento_Min_2m,Rango_Viento_2m
0,2020,1,USAQUEN,4.73197,-74.05163,4653292,213354,10.905097,212731,215097,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
1,2020,1,CHAPINERO,4.67040,-74.04338,2281601,93362,12.219109,93016,93684,...,89.65,94.8,2.310,6.405,0.140,3.07,1.395,4.175,0.065,2.085
2,2020,1,SANTA FE,4.60947,-74.08135,843631,33789,12.483811,33524,34703,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
3,2020,1,SAN CRISTOBAL,4.58843,-74.07080,1771678,80676,10.980205,80311,81072,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945
4,2020,1,USME,4.53286,-74.11451,1570282,78240,10.035033,77876,79159,...,78.20,78.8,2.275,6.650,0.205,2.31,1.430,4.735,0.150,1.945


In [178]:
ruta1 = r"C:\Users\Xiomara\Documents\Projects\Tesis\cd_tesis\Par_meteo\processed_db\\base_pbi.xlsx"
df_PBI.to_excel(ruta1, index=False)